### PR TITLE
resolves #2035: refactor lucene index auto complete and spell check

### LINKED ISSUE #2035: refactor lucene index auto complete and spell check
Lucene's auto-complete and spell check are functional, however, cause problems when combined with e.g. synthetic record types, as the implemented code-path violates certain assumptions. Auto-complete for instance fetches the base record in the cursor returned by the index maintainer. This issue should correct these issues.

### PR DESCRIPTION
**Pull request by normen662**

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/AvailableFields.java [hunk 1]
```diff
@@ -109,17 +109,92 @@ public IndexKeyValueToPartialRecord.Builder buildIndexKeyValueToPartialRecord(@N
     /**
      * Get the fields that are available from just the index scan part of an index plan.
      */
-    public static AvailableFields fromIndex(@Nonnull RecordType recordType,
-                                            @Nonnull Index index,
-                                            @Nonnull PlannableIndexTypes indexTypes,
-                                            @Nullable KeyExpression commonPrimaryKey,
-                                            @Nonnull RecordQueryPlanWithIndex indexPlan) {
+    public static AvailableFields fromIndex(@Nonnull final RecordType recordType,
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/AvailableFields.java:112]: This file mostly contains refactorings to allow the covering index code path for Lucene to directly translate the primary key information.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryIndexPlan.java [hunk 2]
```diff
@@ -326,15 +325,7 @@ private <M extends Message> RecordCursor<IndexEntry> executeEntriesWithOverScan(
...
-
-    @Nonnull
-    @Override
-    public <M extends Message> RecordCursor<FDBQueriedRecord<M>> fetchIndexRecords(@Nonnull final FDBRecordStoreBase<M> store,
-                                                                                   @Nonnull final RecordCursor<IndexEntry> entryRecordCursor,
-                                                                                   @Nonnull final ExecuteProperties executeProperties) {
-        return fetchIndexRecords.fetchIndexRecords(store, entryRecordCursor, executeProperties);
-    }
-
+
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryIndexPlan.java:328]: This for some reason contained the default implementation.

**Review by scottfines — DISMISSED** [review]

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexQueryPlan.java [hunk 3]
```diff
@@ -68,82 +62,28 @@ public class LuceneIndexQueryPlan extends RecordQueryIndexPlan implements PlanWi
...
+    public <M extends Message> RecordCursor<FDBQueriedRecord<M>> fetchIndexRecords(@Nonnull final FDBRecordStoreBase<M> store,
+                                                                                   @Nonnull final EvaluationContext evaluationContext,
+                                                                                   @Nonnull final Function<byte[], RecordCursor<IndexEntry>> entryCursorFunction,
+                                                                                   @Nullable final byte[] continuation,
+                                                                                   @Nonnull final ExecuteProperties executeProperties) {
         final RecordMetaData metaData = store.getRecordMetaData();
         final Index index = metaData.getIndex(indexName);
         final Collection<RecordType> recordTypes = metaData.recordTypesForIndex(index);
         if (recordTypes.size() != 1) {
             throw new RecordCoreException("No lucene index should span multiple record types");
```
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexQueryPlan.java:84]: If this error check is all that's left of the override, can it be moved from runtime into the planner and the override done away with altogether? Or pulled out into a method used for effect here and for value in the spell-checker's override?

**Review by MMcM — CHANGES_REQUESTED** [review]

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexQueryPlan.java (hunk 3 above)
Comment (normen662) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexQueryPlan.java:84]: good point -- will see if I can improve that

Comment (normen662) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexQueryPlan.java]: done

**Review by MMcM — APPROVED** [review]

**Pull request merged**